### DISCUSSION TITLE
[API Design] Link with separate getURL/getMappedURL/getExternalizedURL methods

### DISCUSSION THREAD

**Discussion by stefanseifert** [adobe/aem-core-wcm-components#1857] [Ideas]
i find it a bit confusing what's to do with those three methods. the javadocs are clear, but I still suppose most users who "just want a link" will be puzzled which is the correct one for their HTL script, and in a big project with lots of developers every developer will use a different method leading to much inconsistency.

in wcm.io Link handler/URL handler for comparison we have also concepts for influencing if links should be externalized or not (avoiding the ill-designed com.day.cq.commons.Externalizer altogether), but this are only optional aspects which can be activated during link resolution with special builder methods, and there is always a sensible default handling which is the right thing to use in 95% of the cases.

providing access to these special case methods in this three variants of URLs may lead to confusion and inconsistencies.

---

this discussion is taken over from the [mailing list](https://groups.google.com/d/msgid/aem-core-components-dev/AS8PR07MB7238137D48504B293171B7F1B8C89%40AS8PR07MB7238.eurprd07.prod.outlook.com) where it did not gather any attention (except from @ky940819).

**Comment by vladbailescu** [comment]
Using a builder pattern would have probably been a better choice than the https://github.com/adobe/aem-core-wcm-components/blob/master/bundles/core/src/main/java/com/adobe/cq/wcm/core/components/services/link/PathProcessor.java. I'm uneasy to deprecate the `PathProcessor` approach but we could try to see if we find a way to improve things while keeping that as well.
